### PR TITLE
Add box-shadow rendering

### PR DESCRIPTION
**Pull request by eKoopmans**
## Feature: Box-shadows

You know 'em, you love 'em: `box-shadow`s! Requests for this feature date back to May 2013.

feature-box-shadow

### Implementation

In `paintElement`, after rendering backgrounds but before borders, there is now `renderShadows`. It uses a new `.mask()` tool which behaves the opposite of `.clip()`, meaning the shadow will be drawn only outside the element container.

This feature includes support of multiple `box-shadow`s.

### Related issues/pull requests

#220, #221, #246, #510, #597, #673, #763

### CODE REVIEW

**Comment by AvanishKumar008** [comment]
@eKoopmans @balupton @gatapia @CyberShadow @andyedinborough  Have any solution for it. I have same issue please help me ASAP.

**Comment by CyberShadow** [comment]
@AvanishKumar008: Don't be a jerk. Mass-noticing people and begging for help is extremely rude. Not to mention that my only relation to this project is a contribution I made in 2012 (yes, over 5 years ago).

**Comment by AvanishKumar008** [comment]
@CyberShadow I am just asking. I am not forcing you. If you want to fix you can.

**Comment by eKoopmans** [comment]
Hi @AvanishKumar008, this is a pull request which means I have solved the problem and I'm asking to add the feature into the repository. It won't be added in until the repo owner approves it.

If you want to use `html2canvas` with this new feature added, you can download [this custom build](https://github.com/eKoopmans/html2canvas/tree/develop/dist) I have made that includes drop-shadow support along with a few other new features and bugfixes. If you find any problems with how drop-shadows are rendered when you use that build, you can comment here to let me know. Thanks!

**Comment by AvanishKumar008** [comment]
@eKoopmans You did great job for box shadow and some other thing as well but full picture not showing. Please check again and fix it.

Best of luck

**Comment by eKoopmans** [comment]
@AvanishKumar008 I'm happy to take a look if you give me a link to an example, but as it is your comment isn't helpful - in all my tests, box-shadows are now rendering correctly.

If your comment is about [rendering of off-screen elements](https://github.com/niklasvh/html2canvas/pull/1083), please comment over there in the appropriate pull request - and again, it's only helpful if you can give me an example of what's not working.

**Comment by joneshwangoo** [comment]
@eKoopmans Thanks a lot. You solved my problem.

**Comment by SteKama** [comment]
hi to everyone..thanks to eKoopmans that built "this custom build"..but i have a problem with 
filter: drop-shadow(5px 5px 5px #222); 
i lost it when generate a canvas..how can i do? (sorry for my english language)

**Comment by eKoopmans** [comment]
Hi @SteKama, really sorry about the long wait! Unfortunately the `filter` CSS is completely separate and isn't supported right now, see #246 #493. Good luck!

**Comment by bastienlemaitre** [comment]
Any update for this PR ?

**Comment by eKoopmans** [comment]
Not yet merged. @niklasvh, I believe most of the logic should still stand up, but I'm not sure about the particulars of how you've changed clip, renderBorder, etc. Basically the functions I added (mask, renderShadows, and shadow) mirrored the behaviour of the existing clip, renderBorders, and rectangle/circle/etc.

**Comment by niklasvh** [comment]
Shouldn't be too hard to integrate this with the rewrite, but I'd rather focus on getting bugfixes done and 1.0.0 out asap before adding any new features.

**Comment by Webictbyleo** [comment]
Not rendering when target element is positioned off screen. Canvas appears blank

**Comment by alex-place** [comment]
Any update on this being merged?

**Comment by CPIEmojiMaker** [comment]
When is it going to be merged? Box-shadows are highly used in my project and it would be amazing if it worked someday... @niklasvh

**Comment by Vincedream** [comment]
emmmm，when will this feature merged？it is a very practical feature!!!

**Comment by skullface** [comment]
+1 to merging this one 🙏

**Comment by qbob** [comment]
+1 to merging this one 🙏

**Comment by liuchouchou0728** [comment]
+1 to merging this one

File: src/renderers/canvas.js [hunk 1]
```diff
@@ -41,6 +41,38 @@ CanvasRenderer.prototype.circleStroke = function(left, top, size, color, stroke,
...
+        }
+        for (var i=0; i<info.px.length; i++) {
+            info[pxPropertyNames[i]] = parseInt(info.px[i]);
+        }
+        return info;
+    };
+    var drawShadow = function(shadow) {
+        var info = parseShadow(shadow);
+        if (!info.inset) {
+            context.shadowOffsetX = info.x;
```
Comment (NetroScript) [inline src/renderers/canvas.js]:
```suggestion
            if (info.spread > 0) {
                for (var i = -1; i < 2; i += 2) {
                    for (var i2 = 1; i2 > -2; i2 -= 2) {
                        context.shadowOffsetX = info.x + i * info.spread;
                        context.shadowOffsetY = info.y + i2 * info.spread;
                        context.shadowColor = info.color;
                        context.shadowBlur = info.blur;
                        context.fill();
                    }
                }
                return;
            }
            context.shadowOffsetX = info.x;
```
When the spread is greater 0 you can draw the shape 4 times to simulate the spread. As long as the spread value is not bigger than double of the actual width / height it should work as intended (or better than not handling it at all).
But it is not possible to do negative spread with this method,

**Comment by hakimio** [comment]
@niklasvh Would you be willing to merge this if I'd integrated it with v1.0.0 and your latest PR ("Typescript conversion")?

**Comment by Charlie85270** [comment]
Any update on this being merged?

**Comment by hakimio** [comment]
@Charlie85270 PR #1848 was merged. It works to some degree.

**Comment by Muneem** [comment]
its still not working for me

**Comment by tonyChenHey** [comment]
* html2canvas 1.0.0-rc.7 <https://html2canvas.hertzen.com>
* Copyright (c) 2020 Niklas von Hertzen <https://hertzen.com>

此最新版本box-shadow的渲染還是有問題
image

**Pull request closed without merging**